### PR TITLE
GG-6759

### PR DESCRIPTION
**Pull request by martin-armstrong**

### CODE REVIEW

**Review by arunankannan — APPROVED** [review]

**Review by arunankannan — DISMISSED** [review]

File: app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    case WhichToUse.New => verifiedHashedEmailRepo.find(mixedCaseEmail.toLowerCase)
+  }
+
+  /**
+   * older plain text collection is stored mixed case, new hashed collection is stored lower cased.
+   */
+  def insert(mixedCaseEmail: String): Future[Unit] = appConfig.verifiedEmailUpdateCollection match {
+    case WhichToUse.Both => verifiedHashedEmailRepo.insert(mixedCaseEmail.toLowerCase)
+      .flatMap(_ => verifiedEmailRepo.insert(mixedCaseEmail))
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala]: I think we should insert first into the old repo, and if that succeeds, then insert into the new repo ... to keep them in-sync

File: app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    case WhichToUse.New => verifiedHashedEmailRepo.insert(mixedCaseEmail.toLowerCase)
+    case WhichToUse.Old => throw new RuntimeException("'old' should be logically impossible here, only 'both' or 'new' supported in config")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala]: Maybe add at the beginning of the string "Post-migration, ".

Also, before migration, could this be set to 'old' here?  As we'll probably go to Prod with that, then enable migration via a timed recycle?

File: conf/application.conf [hunk 3]
```diff
@@ -64,6 +65,29 @@ maxDifferentEmails = 10
 
 passcodeExpiryMinutes = 15
 
+verifiedEmailRepo {
+    hashKey = somehashkey
+    replaceIndex = false # If true rebuilds indexes on startup
+    ttlDays = 2555 # 2555 days = 7 years
```
Comment (tim-holbrook) [inline conf/application.conf:71]: Agreed, 7 years feels like a good default.  Let's check with the PO he's ok with this.

File: conf/application.conf [hunk 4]
```diff
@@ -64,6 +65,29 @@ maxDifferentEmails = 10
...
+    enabled = false
+    startAfterMinutes = 1
+    batchSize = 1000
+    batchDelayMillis = 500
+    maxDurationSeconds = 3600
+}
+
 controllers {
     # 300 is the default, you may need to change this according to your needs
     confidenceLevel = 300
```
Comment (tim-holbrook) [inline conf/application.conf]: 300 => 250

File: conf/application.conf [hunk 5]
```diff
@@ -64,6 +65,29 @@ maxDifferentEmails = 10
...
+// set to both initially, then switch to new once migration completed (i.e. ready to remove the old collection)
+verifiedEmailUpdateCollection = "both"
+
+#GG-6759 remove following email -> hashed email collection migration
+emailMigration {
+    enabled = false
+    startAfterMinutes = 1
+    batchSize = 1000
+    batchDelayMillis = 500
+    maxDurationSeconds = 3600
```
Comment (tim-holbrook) [inline conf/application.conf:88]: Let's check the maths together, inc. maxDuration, as it'll be a huge collection

File: it/uk/gov/hmrc/emailverification/JourneyWireMockSpec.scala [hunk 6]
```diff
@@ -551,14 +561,14 @@ class JourneyWireMockSpec extends BaseISpec with Injecting {
       )
     }
 
-    def verifyEmailVerificationOutcomeEventFired(times: Int = 1, emailsJsonArray:String, expectedStatus:Int): Unit = {
+    def verifyEmailVerificationOutcomeEventFired(times: Int = 1, emailJson:String, expectedStatus:Int): Unit = {
       Thread.sleep(100)
```
Comment (tim-holbrook) [inline it/uk/gov/hmrc/emailverification/JourneyWireMockSpec.scala]: Can we use patience or eventually (defaults seem ok), vs sleep ?

File: conf/application.conf (hunk 5 above)
Comment (martin-armstrong) [reply conf/application.conf:88]: Yep currently 2.264 million verified emails in the collection. If it takes 1 second (including batch delay) per 1000 records, thats 2264 seconds or 37.66 minutes, so I've got max duration cut off at 1 hour.

File: app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala (hunk 1 above)
Comment (martin-armstrong) [reply app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala]: 👍

File: conf/application.conf (hunk 3 above)
Comment (martin-armstrong) [reply conf/application.conf:71]: 👍

File: conf/application.conf (hunk 4 above)
Comment (martin-armstrong) [reply conf/application.conf]: Change this but I've a feeling this is legacy bootstrap config, our code doesn't actually read it

File: it/uk/gov/hmrc/emailverification/JourneyWireMockSpec.scala (hunk 6 above)
Comment (martin-armstrong) [reply it/uk/gov/hmrc/emailverification/JourneyWireMockSpec.scala]: 👍

**Review by tim-holbrook — APPROVED** [review]

**Pull request merged**